### ISSUE TITLE
[Bug]The file copied by TF from HDFS to local may be wrong, when HDFS file is being overwritten

### ISSUE THREAD

**Issue by yuanbopeng** [tensorflow/tensorflow#42597]
This is a issue from TaiJi AI platform in Tencent.

**System information**
- OS Platform and Distribution : Linux version 4.14.105-1-tlinux3-0010
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.13.1（we use）and the latest version also has this problem
- Python version: 3.6
- C++ version: 11

**Describe the current behavior**
Our training sample data is generated by the spark program and stored on HDFS, an example of a training sample file: `hdfs://xxxx/example/20200822/part-r-0000036.tfr.gz`, and the file data is compressed by gzip.
The trigger condition of the training program is that the `_SUCCESS` file appears under `hdfs://xxxx/example/20200822/`. The training program first downloads the training samples on HDFS to the local, and then reads the local data for training. When the training program and the spark program are running at the same time, the downloaded HDFS file may be overwritten by the spark program, causing the gzip file downloaded to the local to be damaged. Once the gzip file is wrong, our tensorflow training program will always stay unzipped, and the CPU utilization rate is high. 
The wrong local gzip file is composed of part of the data of the HDFS file before and after overwriting.

code:
```
 auto env = tensorflow::Env::Default();
 auto st = env->CopyFile(src_file, des_file);
```
process pstack info：
![image](https://user-images.githubusercontent.com/70072713/90976846-30be5080-e573-11ea-9f02-dace76b15584.png)
top info:
![image](https://user-images.githubusercontent.com/70072713/90976850-3fa50300-e573-11ea-90bf-56ccf06172c7.png)

**Describe the expected behavior**
The local gzip file is consistent with the data of the HDFS file before overwriting, or the data of the HDFS file after overwriting, instead of containing the data of the HDFS file before and after overwriting

**issues analysis**
In order to solve the [issue:5438](https://github.com/tensorflow/tensorflow/issues/5438) that the tensorboard needs to get the latest data written, the HDFS file is reopened in [the HDFSRandomAccessFile Read](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L226): when n>0 r=0, call hdfsOpenFile to reopen the HDFS file. Please refer to this [commit](https://github.com/tensorflow/tensorflow/commit/e6e8d8715552d8890c0dd10f49ec3dff931a9926) for details.
Before calling hdfsOpenFile, if the HDFS file is overwritten, a new HDFS file is generated.
After calling hdfsOpenFile, it will point to the new HDFS file. If the size of the new HDFS file is larger than the size of the old HDFS file, the HDFS file copied to the local file system by FileSystemCopyFile contains part of the data of the new and old HDFS files, causing the local gzip file to be wrong

**temp solution**: [patch-1](https://github.com/tensorflow/tensorflow/pull/42598)
[FileSystemCopyFile](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/env.cc#L466) avoids triggering the hdfsOpenFile operation of the HDFSRandomAccessFile Read. The size of the file copy is based on the file size, not based on kCopyFileBufferSize. The implementation principle of the temporary solution is the same as that of [ReadFileToString](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/env.cc#L423), but it is still possible that the file copied to the local file is wrong. Because `GetFileSize` and `READ` cannot form an atomic operation. For example, when the file size is obtained through GetFileSize, the HDFS file is overwritten, and the data of the new file is read based on the size of the old file. However, the possibility that the local file of the temporary solution is wrong is far less than the original solution. Generally speaking, reading the file data to the end of the file is a time-consuming operation, and the time-consuming operation of obtaining the file size is negligible

**may be a better solution**: [patch-2](https://github.com/tensorflow/tensorflow/pull/42599)
I'm not sure if this solution is a better solution. In some scenarios I don't know, it may require further discussion. RandomAccessFile READ is an abstraction of the operations supported by each file system, and the specific implementation is transparent to users. Adding the `hdfsOpenFile` to the [HDFSRandomAccessFile READ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L226) to read the latest data is a hidden and dangerous behavior. Because hdfsOpenFile may point to new files, data inconsistencies may occur. More fatally, there are a large number of methods that depend on the READ, which may cause some behaviors that are not what we expect, which is the root of all errors. I think it is better for users to use READ and REOPEN to obtain the latest data in the program.

**accepted solution**: [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)
Quoting mihaimaruseac's comment:
```
Patch-1 has the issue of breaking separation of concern design principles (what happens if there is a new scheme for hdfs? We would have a bug in there until someone remembers the additional if). 
Patch-2 has the issue of removing a test that was added for creating a bug.
```
In order to overcome the shortcomings of patch-1 and patch-2, a switch is added to [patch-3](https://github.com/tensorflow/tensorflow/pull/42860),  the default HDFS_DISABLE_READ_EOF_RETRIED is false. [patch-3](https://github.com/tensorflow/tensorflow/pull/42860) will not remove the [WriteWhileReading](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system_test.cc#L202) test case, and it can also solve the problem we encountered. If you need to turn off the HDFS_READ_EOF_RETRIED, set the environment variable:
 ```
source HDFS_DISABLE_READ_EOF_RETRIED=1
```
For more details, please refer to the comments below.

**Comment by mihaimaruseac** [comment]
I think this is an issue that can be solved by [the transaction support introduced by the new RFC](https://github.com/tensorflow/community/pull/245)

**Comment by yuanbopeng** [comment]
@mihaimaruseac Thanks for the comment. The reopen operation is added to [HDFSRandomAccessFile READ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L226)  to solve the problem of reading the latest data in the issue. I think that the reopen operation should not be added to [the low level filesystem API](https://github.com/tensorflow/community/blob/9c248b5603256c0641a05dd202adb2942c8af500/rfcs/20190506-filesystem-plugin-modular-tensorflow.md#low-level-filesystem-api), and it should be consistent with the hdfs file system. Because this has destroyed the semantics of read provided by hdfs, read reads from the current file name handle, not from a specific file name. If the file is just appended, this will work very well, like adding a retry in the read, reacquiring a new file handle to get the latest data. However, the file is overwritten, and the data read may be wrong.
![image](https://user-images.githubusercontent.com/70072713/91115229-573bd300-e6bc-11ea-8c9f-04ed27005a9e.png)

**Comment by yuanbopeng** [comment]
One more. My understanding, about [the transaction support introduced by the new RFC](https://github.com/tensorflow/community/pull/245) I have read.
As [vnno2409 commented](https://github.com/tensorflow/community/pull/245#issuecomment-628380634), HDFS does not support transactions, and transaction functions require file system support, such as s3 or gcs. Therefore, to less intrusive, [the transaction-related APIs on HDFS](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc) only pass the transaction parameters, and there is no transaction implementation.
Because HDFS does not support transactions, I think the reopening method is used in [the HDFS read method](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L255) to support simultaneous reading and writing scenarios, and data accuracy cannot be guaranteed. We should circumvent this implementation method, which undermines the semantics provided by HDFS and causes the read data to be wrong

Finally, because this question is very important to us, I look forward to @mihaimaruseac  your comments, thanks :).

**Comment by mihaimaruseac** [comment]
So the big issue is that the PRs remove an existing test that was added for a different issue. This is not good, we should not break existing behavior if possible.

From what I gather, you want to support the scenario where you started reading from a file but it got overwritten. This is a classic TOCTOU scenario, not even POSIX handles it by itself:

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main() {
	FILE *f = fopen("file", "r");
	char buffer[10];
	fgets(buffer, 10, f);
	printf("Got \"%s\"\n", buffer);
	sleep(10);
	fgets(buffer, 10, f);
	printf("Got \"%s\"\n", buffer);
	return 0;
}
```

```console
[mm] λ gcc -Wall -Wextra test.c -o ./test
[mm] λ echo "asdf" > file
[mm] λ ./test 
Got "asdf
"
Got "asdf
"
```

vs with overwriting the file during the 10s sleep:

```console
[mm] λ ./test 
Got "asdf
"
Got "fghijklmn"
```

vs appending to the file

```console
[mm] λ ./test 
Got "asdf
"
Got "abcdefghi"
```

I think that if we want to handle these scenarios we need to add transaction support to the filesystem implementation. Even if HDFS itself does not support transactions, we can add this layer as a middleware on top of it.

**Comment by yuanbopeng** [comment]
@mihaimaruseac 
Ok. I agree with you that to solve this problem and delete the test may cause other issues. Therefore, [patch-1](https://github.com/tensorflow/tensorflow/pull/42598) may be better. According to your suggestionIn, in the HDFS scenario, [FileSystemCopyFile](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/env.cc#L466) adapted to avoid triggering [HDFSRandomAccessFile READ reopen](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L255).

In addition, I made some modifications and tests based on your POSIX test example, and I am not sure whether my conclusion is the same as yours. In POSIX, the additional data can be read, but the overwritten data cannot be read; but in HDFS read with reopen, both the additional data and the overwritten data can be read. The read data is a mixture before and after the overwrite file.

```c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main() {
	FILE *f = fopen("file", "r");
	char buffer[10];
	char *first = fgets(buffer, 10, f);
        printf("Got \"%s\"\n", first);
	sleep(20);
	char *second = fgets(buffer, 10, f);
        printf("Got \"%s\"\n", second);
	return 0;
}
```
```
$ echo "asdf" > file
$ ./test_read
Got "asdf
"
Got "(null)"
```
vs with overwriting the file during the 15s sleep:
```
./test_read
Got "aaa
"
Got "(null)"
```

```
$ echo "aaa" > file
--------during the 20s sleep--------
$ echo "bbb" > file
$ cat file
bbb
```

vs appending to the file during the 15s sleep:
```
./test_read
Got "aaa
"
Got "bbb
"
```
```
$ echo "aaa" > file
--------during the 20s sleep--------
$ echo "bbb" >> file
$ cat file
aaa
bbb
```

**Comment by yuanbopeng** [comment]
@mihaimaruseac 
Since I did not receive your comment, it may be because my previous reply was not clear enough.
Let me add one more thing. I think the biggest problem is that POSIX will not read the overwritten file, but the [HDFS read API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L226) encapsulated by tensorflow does. For details, please refer to the previous comment. [patch-1](https://github.com/tensorflow/tensorflow/issues/42597#issuecomment-682985582) is to solve the issue of copying the file being overwritten. Since the copied HDFS file has a certain size, the file size can be passed in to avoid the bug of reading overwritten files caused by [reopon](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L255).

I look forward to @mihaimaruseac your comments, thanks :).

**Comment by mihaimaruseac** [comment]
We have the exact same test for POSIX behavior with exactly the same results.

POSIX reads from the overwritten file, from the current cursor position.

This should be solved via transaction support, I think. Patch-1 has the issue of breaking separation of concern design principles (what happens if there is a new scheme for hdfs? We would have a bug in there until someone remembers the additional if). Patch-2 has the issue of removing a test that was added for creating a bug.

If anything, I'd prefer patch 2 to patch 1 but I'd prefer adding transactional support instead of either of the two patches.

**Comment by yuanbopeng** [comment]
@mihaimaruseac  
Confirm that the `via transaction support ` here means: to migrate data from HDFS to the transaction file system, and then read and write data from the transaction file system to avoid the problem of overwritten files?

If this is the case, it may not be suitable for our scenario, because all our training is stored on HDFS, it is almost impossible for us to switch other transaction file systems. Moreover, our data volume is very huge, so storing it in HDFS is a better choice.

Reading EOF will try to open the file again. This is a serious bug in our scenario.
How about [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)? patch-3 will add a switch, the default DISABLE_HDFS_READ_EOF_RETRIED is false. if you need to turn off the HDFS_READ_EOF_RETRIED function, set DISABLE_HDFS_READ_EOF_RETRIED=1; patch-3 will not destroy the original test case, and it can also solve the problem we encountered.

Can you help review the [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)? I look forward to @mihaimaruseac your comments, thanks :).

**Comment by yuanbopeng** [comment]
@mihaimaruseac 
Update [patch-3](), refer to the naming of [S3_DISABLE_MULTI_PART_DOWNLOAD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/s3/s3_file_system.cc#L482), `DISABLE_HDFS_READ_EOF_RETRIED ` was renamed to `HDFS_DISABLE_READ_EOF_RETRIED`.

Can you help review [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)? I look forward to @mihaimaruseac  your comments, thanks :).

**Comment by mihaimaruseac** [comment]
> @mihaimaruseac
> I want to confirm that the `via transaction support ` here means: to migrate data from HDFS to the transaction file system, and then read and write data from the transaction file system to avoid the problem of overwritten files?

I meant submitting a patch to convert the HDFS support to https://github.com/tensorflow/community/pull/245.

> 
> If not, maybe I misunderstood what you mean. Can you show me specific HDFS examples? If it is, it may not be suitable for solving our issue, because all our training data is stored on HDFS, it is almost impossible for us to switch other transaction file systems. And our data volume is very huge, at least petabytes or more, storing in HDFS is a better choice.

You won't need to change your data storage. All I am saying is that the best fix here is to implement transaction support to HDFS layer in TF.

> Can you help review [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)? I look forward to @mihaimaruseac your comments, thanks :).

Looking over the patch now

**Comment by yuanbopeng** [comment]
@mihaimaruseac 
> > Can you help review [patch-3](https://github.com/tensorflow/tensorflow/pull/42860)? I look forward to @mihaimaruseac your comments, thanks :).

Thank you so much for your reply, I figured out your suggestion. Our disagreement may be that I think that hdfs itself has no transactions, so it is impossible to implement transactions on the tensorflow side. I am not sure if you have more material to prove that it is possible to implement transactions on the tensorflow side. If so, can you share it?

**Comment by mihaimaruseac** [comment]
Basically, what you are doing with the retries is a component of transactions. TF could use the transaction token to know when to do the transaction retry, in all operations. A transaction is nothing more than a set of files which either all change at the same time or not.

In the end, it's the same as how TCP protocol is implemented over the IP protocol. TCP is connection oriented, requires handshakes, checksums, has a lot of flow control. IP has none of these, is just best-effort. The TCP implementation contains all the code needed to implement the additional layers. Contrast with UDP which just encapsulates the data and throws it over to the IP layer.

**Comment by yuanbopeng** [comment]
> Basically, what you are doing with the retries is a component of transactions. TF could use the transaction token to know when to do the transaction retry, in all operations. A transaction is nothing more than a set of files which either all change at the same time or not.

I have a doubt whether this implementation requires a central service to store the transaction token

**Comment by mihaimaruseac** [comment]
This has been solved by the third patch
